NoSoft-SA/nspack
# Manage pallet exiting from stock or reversing back into stock

Summary of this change:
All processes that set or unset the `exit_ref` of a pallet have been changed to use this central service that incorporates:

* Update the pallet and sequences (when applicable) with the appropriate exit ref
* Call the MoveStock service to set the new location
* Apply the status to pallet(s) and sequences

The calculation of the `units_in_location` on the locations table has been improved to be accurate.

CHANGELOG entry:

### Changed
- Handle exit ref changes centrally and improve logging of `units_in_location`.

**How to test this code via the UI:**

Run the scriptfix `RACK_ENV=production ruby scripts/base_script.rb ReCalculateUnitsInLocationAndShipBins`

_Perform the following actions:_
* Reworks scrap a pallet
* Reworks scrap several pallets
* Reworks unscrap a pallet
* Reworks unscrap several pallet
* Reworks repack a pallet
* Resorks restore a repacked pallet
* Buildup a pallet where the last sequence is removed from a pallet
* Carton palletizing - transfer the last carton of a pallet to another bay
* Convert carton to bin (scraps the converted pallet)
* Convert pallet to rebin (scraps the converted pallet)
* Ship a load
* Unship a load

_Check the following after each change:_
* pallets.scrapped
* pallets.scrapped_at
* pallets.exit_ref
* pallets.status
* pallets.location
* Serialized stock movement for the pallet
* pallet sequences.scrapped_at
* pallet_sequences.pallet_id
* pallet_sequences.scrapped_from_pallet_id
* pallet_sequences.exit_ref
* pallet_sequences.status
* locations.units_in_location

To check that the units in location are accurately updated, run this query which should return 0 rows:
```sql
      SELECT id, location_long_code, units_in_location,
        ((SELECT COUNT(*) FROM pallets WHERE location_id = locations.id) + (SELECT COUNT(*) FROM rmt_bins WHERE location_id = locations.id)) AS all_units,
        (SELECT COUNT(*) FROM rmt_bins WHERE location_id = locations.id) AS bins,
        (SELECT COUNT(*) FROM pallets WHERE location_id = locations.id) AS pallets,
        units_in_location - ((SELECT COUNT(*) FROM pallets WHERE location_id = locations.id) + (SELECT COUNT(*) FROM rmt_bins WHERE location_id = locations.id)) AS diff
      FROM locations
      WHERE units_in_location <> ((SELECT COUNT(*) FROM pallets WHERE location_id = locations.id) + (SELECT COUNT(*) FROM rmt_bins WHERE location_id = locations.id))
      ORDER BY location_long_code;
```
